collector/mssql.go

```diff
@@ -0,0 +1,2757 @@
+			[]string{"instance"},
+			nil,
+		),
+	}, nil
+}
+
+// Collect sends the metric values for each metric
+// to the provided prometheus Metric channel.
+func (c *MSSQLCollector) Collect(ch chan<- prometheus.Metric) error {
+	for instance := range sqlInstances {
+		log.Debugf("mssql collector iterating sql instance %s.", instance)
+
+		// Win32_PerfRawData_{instance}_SQLServerAvailabilityReplica
+		if desc, err := c.collectAvailabilityReplica(ch, instance); err != nil {
+			log.Error("failed collecting MSSQL GeneralStatistics metrics:", desc, err)
+			return err
+		}
+
+		// Win32_PerfRawData_{instance}_SQLServerBufferManager
+		if desc, err := c.collectBufferManager(ch, instance); err != nil {
+			log.Error("failed collecting MSSQL BufferManager metrics:", desc, err)
+			return err
+		}
+
+		// Win32_PerfRawData_{instance}_SQLServerDatabaseReplica
+		if desc, err := c.collectDatabaseReplica(ch, instance); err != nil {
+			log.Error("failed collecting MSSQL DatabaseReplica metrics:", desc, err)
+			return err
+		}
+
+		// Win32_PerfRawData_{instance}_SQLServerDatabases
+		if desc, err := c.collectDatabases(ch, instance); err != nil {
+			log.Error("failed collecting MSSQL Databases metrics:", desc, err)
+			return err
+		}
+
+		// Win32_PerfRawData_{instance}_SQLServerGeneralStatistics
+		if desc, err := c.collectGeneralStatistics(ch, instance); err != nil {
+			log.Error("failed collecting MSSQL GeneralStatistics metrics:", desc, instance, err)
+			return err
+		}
+
+		// Win32_PerfRawData_{instance}_SQLServerLocks
+		if desc, err := c.collectLocks(ch, instance); err != nil {
+			log.Error("failed collecting MSSQL Locks metrics:", desc, err)
+			return err
+		}
+
+		// Win32_PerfRawData_{instance}_SQLServerMemoryManager
+		if desc, err := c.collectMemoryManager(ch, instance); err != nil {
+			log.Error("failed collecting MSSQL MemoryManager metrics:", desc, err)
+			return err
+		}
+
+		// Win32_PerfRawData_{instance}_SQLServerSQLStatistics
+		if desc, err := c.collectSQLStats(ch, instance); err != nil {
+			log.Error("failed collecting MSSQL SQLStats metrics:", desc, err)
+			return err
+		}
+	}
+
+	return nil
+}
+
+type win32PerfRawDataSQLServerAvailabilityReplica struct {
+	BytesReceivedfromReplicaPersec uint64
+	BytesSenttoReplicaPersec       uint64
+	BytesSenttoTransportPersec     uint64
+	FlowControlPersec              uint64
+	FlowControlTimemsPersec        uint64
+	ReceivesfromReplicaPersec      uint64
+	ResentMessagesPersec           uint64
+	SendstoReplicaPersec           uint64
+	SendstoTransportPersec         uint64
+}
+
+func (c *MSSQLCollector) collectAvailabilityReplica(ch chan<- prometheus.Metric, sqlInstance string) (*prometheus.Desc, error) {
+	var dst []win32PerfRawDataSQLServerAvailabilityReplica
+	class := fmt.Sprintf("Win32_PerfRawData_%s_SQLServerAvailabilityReplica", sqlInstance)
+	q := queryAllForClass(&dst, class)
+	if err := wmi.Query(q, &dst); err != nil {
+		return nil, err
+	}
+
+	if len(dst) > 0 {
```

Comment:
classes [AvailabilityReplica, DatabaseReplica, Databases, Locks] return multiple results.  
those results should be iterated over and each metric recorded and tagged.